core/desugarFuns.ml

```diff
@@ -67,19 +67,30 @@ class desugar_funs env =
 object (o : 'self_type)
   inherit (TransformSugar.transform env) as super
 
+  method private desugarFunLit argss lin lam location tvs =
+    let inner_mb     = snd (last argss) in
+    let (o, lam, rt) = o#funlit inner_mb lam in
+    let ft = List.fold_right (fun (args, mb) rt -> `Function (args, mb, rt))
+                             argss rt in
+    let f = gensym ~prefix:"_fun_" () in
+    let body, tvs, ft = match tvs with
+      | [] -> Var f, [], ft
+      | _  ->
+         let (tvs, tyargs), ft = Generalise.generalise env.Types.var_env ft in
+         let ft = Instantiate.freshen_quantifiers ft in
+         tappl (Var f, tyargs), tvs, ft in
+    let (bndr, lin, tvs, loc, ty) =
+      unwrap_def (binder ~ty:ft f, lin, (tvs, lam), location, None) in
+    let e = block_node ([with_dummy_pos (Fun (bndr, lin, tvs, loc, ty))],
+                         with_dummy_pos body)
+    in (o, e, ft)
+
   method! phrasenode : Sugartypes.phrasenode -> ('self_type * Sugartypes.phrasenode * Types.datatype) = function
+    | TAbstr (tvs , {node = FunLit (Some argss, lin, lam, location); _ }) ->
```

Comment:
Can `TAbstr`s be nested?  (Or are we completely certain that they won't ever be?)

I can imagine that in an arbitrary `Sugartypes.phrasenode` that there might be an outer `TAbstr`, then something else, then another `TAbstr`, then a `FunLit` (maybe with more other stuff in between).  So it seems like in the general case we should keep track of all of the (distinct?) `tvs` encountered so far and pass them all into `o#desgarFunLit` when we encounter a `FunLit`.  Otherwise we will get mysterious match failures and/or missing type variables.

Reply:
The intention here is to catch the `TAbstr` inserted by typeSugar, because that `TAbstr` contains tyvars that are abstract in the `FunLit`. Here's the relevant code from `typeSugar.ml`:

```OCaml
let e = FunLit (Some argss, lin, (List.map (List.map erase_pat) pats, erase body), location) in
  if Settings.get_value Instantiate.quantified_instantiation then
    let (qs, _tyargs), ftype = Utils.generalise context.var_env ftype in
    let _, ftype = Instantiate.typ ftype in
      tabstr (qs, e), ftype, StringMap.filter (fun v _ -> not (List.mem v vs)) (usages body)
  else
    e, ftype, StringMap.filter (fun v _ -> not (List.mem v vs)) (usages body)
```

Reply:
Hmm.  So I guess this relates to what we just discussed regarding whether the closing should happen in `typeSugar` or a separate desugaring pass.  Now that I've looked at this I lean towards the former - would that be possible or is there some obstacle?  It would be a good invariant that after `typeSugar` does its thing, the resulting AST is still well-typed.

Reply:
To be able to do this in typeSugar we would have to desugar anonymous functions there, because type application needs to go inside the block currently generated in desugarFuns. I agree that "resulting AST is still well-typed" is a perfectly reasonable invariant for a typechecker :-) But I think that in this case it is typeSugar that does more than it actually should, i.e. it should just type check the existing AST, rather than also transform it.